File: app/src/main/java/com/stupidtree/cloudliter/ui/group/GroupEditorViewModel.java
package com.stupidtree.cloudliter.ui.group;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.stupidtree.cloudliter.data.model.RelationGroup;
import com.stupidtree.cloudliter.data.model.UserLocal;
import com.stupidtree.cloudliter.data.repository.GroupRepository;
import com.stupidtree.cloudliter.data.repository.LocalUserRepository;
import com.stupidtree.cloudliter.ui.base.DataState;
import com.stupidtree.cloudliter.ui.base.Trigger;
import com.stupidtree.cloudliter.ui.myprofile.ChangeInfoTrigger;

import java.util.List;
import java.util.Objects;

public class GroupEditorViewModel extends ViewModel {
    /**
     * 数据区
     */
    //数据本体
    LiveData<DataState<List<RelationGroup>>> listData;
    //Trigger:控制↑的刷新
    MutableLiveData<Trigger> listDataController = new MutableLiveData<>();

    //状态数据：添加分组的结果
    LiveData<DataState<String>> addGroupResult;
    //Trigger：控制添加请求的发送
    MutableLiveData<ChangeInfoTrigger> addGroupController = new MutableLiveData<>();

    //状态数据：删除分组的结果
    LiveData<DataState<String>> deleteGroupResult;
    //Trigger：控制删除请求的发送
    MutableLiveData<ChangeInfoTrigger> deleteGroupController = new MutableLiveData<>();
    /**
     * 仓库区
     */
    //好友分组
    GroupRepository groupRepository;
    //本地用户
    LocalUserRepository localUserRepository;


    public GroupEditorViewModel() {
        groupRepository = GroupRepository.getInstance();
        localUserRepository = LocalUserRepository.getInstance();
    }

    public LiveData<DataState<List<RelationGroup>>> getListData() {
        if (listData == null) {
            listData = Transformations.switchMap(listDataController, input -> {
                if (input.isActioning()) {
                    UserLocal userLocal = localUserRepository.getLoggedInUser();
                    if (userLocal.isValid()) {
                        return groupRepository.queryMyGroups(userLocal.getToken());
                    } else {
                        return new MutableLiveData<>(new DataState<>(DataState.STATE.NOT_LOGGED_IN))
                                ;
                    }

                }
                return new MutableLiveData<>(new DataState<>(DataState.STATE.NOTHING));
            });
        }
        return listData;
    }

    public LiveData<DataState<String>> getAddGroupResult() {
        if(addGroupResult==null){
            //也是一样的
            addGroupResult = Transformations.switchMap(addGroupController, input -> {
                if(input.isActioning()){
                    UserLocal userLocal = localUserRepository.getLoggedInUser();
                    if(userLocal.isValid()){
                        //System.out.println("viewmodel stage:func: getAddGroupResult return is "+groupRepository.addMyGroups(userLocal.getToken(),input.getValue()));
                        return groupRepository.addMyGroups(Objects.requireNonNull(userLocal.getToken()),input.getValue());
                    }else{
                        //System.out.println("viewmodel stage: no log in,func: getAddGroupResult return is "+groupRepository.addMyGroups(userLocal.getToken(),input.getValue()));
                        return new MutableLiveData<>(new DataState<>(DataState.STATE.NOT_LOGGED_IN));
                    }
                }
                return new MutableLiveData<>();
            });
        }
        return addGroupResult;
    }

    public LiveData<DataState<String>> getDeleteGroupResult() {
        if(deleteGroupResult==null){
            //也是一样的
            deleteGroupResult = Transformations.switchMap(deleteGroupController, input -> {
                if(input.isActioning()){
                    UserLocal userLocal = localUserRepository.getLoggedInUser();
                    if(userLocal.isValid()){
                        //System.out.println("viewmodel stage:func: getDeleteGroupResult return is "+groupRepository.deleteMyGroups(userLocal.getToken(),input.getValue()));
                        return groupRepository.deleteMyGroups(Objects.requireNonNull(userLocal.getToken()),input.getValue());
                    }else{
                        //System.out.println("viewmodel stage: no log in,func: getDeleteGroupResult return is "+groupRepository.deleteMyGroups(userLocal.getToken(),input.getValue()));
                        return new MutableLiveData<>(new DataState<>(DataState.STATE.NOT_LOGGED_IN));
                    }
                }
                return new MutableLiveData<>();
            });
        }
        return deleteGroupResult;
    }

    /**
     * 发起添加好友分组请求
     * @param group 新组名
     */
    public void startAddGroup(String group){
        System.out.println("viewmodel stage:func: startAddGroup group is "+group);
        addGroupController.setValue(ChangeInfoTrigger.getActioning(group));
    }

    /**
     * 发起删除好友分组请求
     * @param group 新组名
     */
    public void startDeleteGroup(String group){
        System.out.println("viewmodel stage:func: startDeleteGroup group is "+group);
        deleteGroupController.setValue(ChangeInfoTrigger.getActioning(group));
    }

    /**
     * 开始刷新页面
     */
    public void startRefresh(){
        listDataController.setValue(Trigger.getActioning());
    }
}

File: app/src/main/java/com/stupidtree/cloudliter/ui/group/GroupEditorViewModel.kt
package com.stupidtree.cloudliter.ui.group

import android.app.Application
import androidx.lifecycle.*
import com.stupidtree.cloudliter.data.model.RelationGroup
import com.stupidtree.cloudliter.data.repository.GroupRepository
import com.stupidtree.cloudliter.data.repository.LocalUserRepository
import com.stupidtree.cloudliter.ui.base.DataState
import com.stupidtree.cloudliter.ui.base.StringTrigger
import com.stupidtree.cloudliter.ui.base.Trigger
import java.util.*

class GroupEditorViewModel(application: Application) : AndroidViewModel(application) {
    /**
     * 数据区
     */
    //数据本体
    var listData: LiveData<DataState<List<RelationGroup>?>>? = null
        get() {
            if (field == null) {
                listData = Transformations.switchMap(listDataController) { input: Trigger ->
                    if (input.isActioning) {
                        val userLocal = localUserRepository.getLoggedInUser()
                        if (userLocal.isValid) {
                            return@switchMap groupRepository.queryMyGroups(userLocal.token!!)
                        } else {
                            return@switchMap MutableLiveData(DataState<List<RelationGroup>?>(DataState.STATE.NOT_LOGGED_IN))
                        }
                    }
                    MutableLiveData(DataState<List<RelationGroup>?>(DataState.STATE.NOTHING))
                }
            }
            return field
        }

    //Trigger:控制↑的刷新
    var listDataController = MutableLiveData<Trigger>()

    //状态数据：添加分组的结果
    var addGroupResult: LiveData<DataState<String?>>? = null
        get() {
            if (field == null) {
                //也是一样的
                addGroupResult = Transformations.switchMap(addGroupController) { input: StringTrigger ->
                    if (input.isActioning) {
                        val userLocal = localUserRepository.getLoggedInUser()
                        if (userLocal.isValid) {
                            return@switchMap groupRepository.addMyGroups(userLocal.token!!, input.data!!)
                        } else {
                            return@switchMap MutableLiveData(DataState<String?>(DataState.STATE.NOT_LOGGED_IN))
                        }
                    }
                    MutableLiveData<DataState<String?>>()
                }
            }
            return field!!
        }

    //Trigger：控制添加请求的发送
    var addGroupController = MutableLiveData<StringTrigger>()

    //状态数据：删除分组的结果
    var deleteGroupResult: LiveData<DataState<String?>>? = null
        get() {
            if (field == null) {
                //也是一样的
                deleteGroupResult = Transformations.switchMap(deleteGroupController) { input: StringTrigger ->
                    if (input.isActioning) {
                        val userLocal = localUserRepository.getLoggedInUser()
                        if (userLocal.isValid) {
                            //System.out.println("viewmodel stage:func: getDeleteGroupResult return is "+groupRepository.deleteMyGroups(userLocal.getToken(),input.getValue()));
                            return@switchMap groupRepository.deleteMyGroups(userLocal.token!!, input.data!!)
                        } else {
                            //System.out.println("viewmodel stage: no log in,func: getDeleteGroupResult return is "+groupRepository.deleteMyGroups(userLocal.getToken(),input.getValue()));
                            return@switchMap MutableLiveData(DataState<String?>(DataState.STATE.NOT_LOGGED_IN))
                        }
                    }
                    MutableLiveData<DataState<String?>>()
                }
            }
            return field!!
        }

    //Trigger：控制删除请求的发送
    var deleteGroupController = MutableLiveData<StringTrigger>()


    //状态数据：重命名分组的结果
    var renameGroupResult: LiveData<DataState<String?>>? = null
        get() {
            if (field == null) {
                renameGroupResult = Transformations.switchMap(renameGroupController) { input ->
                    if (input.isActioning) {
                        val userLocal = localUserRepository.getLoggedInUser()
                        if (userLocal.isValid) {
                            //System.out.println("viewmodel stage:func: getrenameGroupResult return is "+groupRepository.renameMyGroups(userLocal.getToken(),input.getValue()));
                            return@switchMap groupRepository.renameMyGroups(userLocal.token!!, input.groupId!!,input.name!!)
                        } else {
                            //System.out.println("viewmodel stage: no log in,func: getrenameGroupResult return is "+groupRepository.renameMyGroups(userLocal.getToken(),input.getValue()));
                            return@switchMap MutableLiveData(DataState<String?>(DataState.STATE.NOT_LOGGED_IN))
                        }
                    }
                    MutableLiveData()
                }
            }
            return field!!
        }

    //Trigger：控制重命名请求的发送
    var renameGroupController = MutableLiveData<RenameGroupTrigger>()


    /**
     * 仓库区
     */
    //好友分组
    var groupRepository: GroupRepository

    //本地用户
    var localUserRepository: LocalUserRepository


    /**
     * 发起添加好友分组请求
     * @param group 新组名
     */
    fun startAddGroup(group: String) {
        addGroupController.value = StringTrigger.getActioning(group)
    }

    /**
     * 发起删除好友分组请求
     * @param group 新组名
     */
    fun startDeleteGroup(group: String) {
        deleteGroupController.value = StringTrigger.getActioning(group)
    }

    /**
     * 发起重命名好友分组请求
     * @param group 新组名
     */
    fun startRenameGroup(groupId: String,newName:String) {
        renameGroupController.value = RenameGroupTrigger.getActioning(groupId,newName)
    }


    /**
     * 开始刷新页面
     */
    fun startRefresh() {
        listDataController.value = Trigger.actioning
    }



    init {
        groupRepository = GroupRepository.instance!!
        localUserRepository = LocalUserRepository.getInstance(application)
    }
}